lib/ex_unit/lib/ex_unit/server.ex

```diff
@@ -84,19 +105,43 @@ defmodule ExUnit.Server do
 
   def handle_call({:add, name, :async}, _from, %{loaded: loaded} = state)
       when is_integer(loaded) do
-    state = update_in(state.async_modules, &[name | &1])
+    state =
+      state.async_modules
+      |> update_in(&[name | &1])
+      |> maybe_add_persistent_async_modules(name)
+
     {:reply, :ok, take_modules(state)}
   end
 
   def handle_call({:add, name, :sync}, _from, %{loaded: loaded} = state)
       when is_integer(loaded) do
-    state = update_in(state.sync_modules, &[name | &1])
+    state =
+      state.sync_modules
+      |> update_in(&[name | &1])
+      |> maybe_add_persistent_sync_modules(name)
+
     {:reply, :ok, state}
   end
 
   def handle_call({:add, _name, _type}, _from, state),
     do: {:reply, :already_running, state}
 
+  defp maybe_add_persistent_async_modules(state, name) do
+    if Application.fetch_env!(:ex_unit, :repeat_until_failure) > 0 do
+      update_in(state.persistent_async_modules, &[name | &1])
+    else
+      state
+    end
+  end
+
+  defp maybe_add_persistent_sync_modules(state, name) do
+    if Application.fetch_env!(:ex_unit, :repeat_until_failure) > 0 do
```

Comment:
Sorry but I am worried about depending on global state here. For example, ExUnit.configure can be called at any time and change the meaning of this. Ideally we will keep most state management in the runner. I can think of two other options:

1. Pass a flag when we call `take_modules` in the runner to also make a copy of the modules we are taking
2. Have the runner collect all modules as it runs them and explicitly push them again on `restore_modules`

I think 2 is the cleanest and requires only a small change to `async_loop`. There is no need of additional state and we most rely on existing server APIs. Thoughts?

Reply:
Thank you for all the feedback, I appreciate that very much. I changed the code to use your suggestion 2.